quic/s2n-quic-qns/src/tls.rs

```diff
@@ -2,9 +2,61 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::Result;
-use std::{path::PathBuf, str::FromStr};
+use s2n_quic::provider::tls::s2n_tls::{keylog::KeyLog, ConfigLoader, ConnectionContext};
+use std::{path::PathBuf, str::FromStr, sync::Arc, time::SystemTime};
 use structopt::StructOpt;
 
+pub static TICKET_KEY: [u8; 16] = [0; 16];
+pub static TICKET_KEY_NAME: &[u8] = "keyname".as_bytes();
+
+pub struct Config {
+    alpns: Vec<String>,
+    certificate: s2n_quic::provider::tls::default::certificate::Certificate,
+    private_key: s2n_quic::provider::tls::default::certificate::PrivateKey,
+}
+
+impl Config {
+    fn build(&self) -> Result<s2n_tls::config::Config, s2n_tls::error::Error> {
+        let mut config_builder = s2n_tls::config::Builder::new();
+        config_builder
+            .enable_session_tickets(true)?
+            .add_session_ticket_key(TICKET_KEY_NAME, &TICKET_KEY, SystemTime::now())?
+            .load_pem(
+                self.certificate.0.as_pem().unwrap(),
+                self.private_key.0.as_pem().unwrap(),
+            )?
+            .set_security_policy(&s2n_tls::security::DEFAULT_TLS13)?
+            .enable_quic()?
+            .set_application_protocol_preference(self.alpns.iter().map(String::as_bytes))?;
+
+        let keylog = KeyLog::try_open();
+        unsafe {
+            if let Some(keylog) = keylog.as_ref() {
+                config_builder
+                    .set_key_log_callback(Some(KeyLog::callback), Arc::as_ptr(keylog) as *mut _)?;
+            }
+        }
+
+        cfg_if::cfg_if! {
+            if #[cfg(all(
+                s2n_quic_unstable,
+                feature = "unstable_client_hello"
+            ))] {
+                use super::unstable::MyClientHelloHandler;
+                config_builder.set_client_hello_callback(MyClientHelloHandler {})?;
+            }
+        }
+
+        config_builder.build()
+    }
+}
+
+impl ConfigLoader for Config {
+    fn load(&mut self, _cx: ConnectionContext) -> s2n_tls::config::Config {
+        Self::build(self).expect("Config builder failed")
```

Comment:
I'd prefer to not have to create a new config on every connection, especially since s2n-tls will scan the system's trust store. You can use the [s2n config as a loader](https://docs.rs/s2n-quic-tls/latest/s2n_quic_tls/trait.ConfigLoader.html#impl-ConfigLoader-for-Config) and it'll just do a cheap clone per connection instead.